EventFilter/HcalRawToDigi/plugins/HcalRawToDigiGPU.cc

```diff
@@ -0,0 +1,206 @@
+  edm::EDPutTokenT<ProductTypef01> digisF01HEToken_;
+  using ProductTypef5 = cms::cuda::Product<hcal::DigiCollection<hcal::Flavor5, hcal::common::ViewStoragePolicy>>;
+  edm::EDPutTokenT<ProductTypef5> digisF5HBToken_;
+  using ProductTypef3 = cms::cuda::Product<hcal::DigiCollection<hcal::Flavor3, hcal::common::ViewStoragePolicy>>;
+  edm::EDPutTokenT<ProductTypef3> digisF3HBToken_;
+
+  cms::cuda::ContextState cudaState_;
+
+  std::vector<int> fedsToUnpack_;
+
+  hcal::raw::ConfigurationParameters config_;
+  // FIXME move this to use raii
+  hcal::raw::InputDataCPU inputCPU_;
+  hcal::raw::InputDataGPU inputGPU_;
+  hcal::raw::OutputDataGPU outputGPU_;
+  hcal::raw::ScratchDataGPU scratchGPU_;
+  hcal::raw::OutputDataCPU outputCPU_;
+};
+
+void HcalRawToDigiGPU::fillDescriptions(edm::ConfigurationDescriptions& confDesc) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("InputLabel", edm::InputTag("rawDataCollector"));
+  auto nFeds = FEDNumbering::MAXHCALuTCAFEDID - FEDNumbering::MINHCALuTCAFEDID + 1;
+  std::vector<int> feds(nFeds);
+  for (int i = 0; i < nFeds; ++i)
+    feds[i] = i + FEDNumbering::MINHCALuTCAFEDID;
+  desc.add<std::vector<int>>("FEDs", feds);
+  desc.add<uint32_t>("maxChannelsF01HE", 10000u);
+  desc.add<uint32_t>("maxChannelsF5HB", 10000u);
+  desc.add<uint32_t>("maxChannelsF3HB", 10000u);
+  desc.add<uint32_t>("nsamplesF01HE", 8);
+  desc.add<uint32_t>("nsamplesF5HB", 8);
+  desc.add<uint32_t>("nsamplesF3HB", 8);
+  desc.add<std::string>("digisLabelF5HB", "f5HBDigisGPU");
+  desc.add<std::string>("digisLabelF01HE", "f01HEDigisGPU");
+  desc.add<std::string>("digisLabelF3HB", "f3HBDigisGPU");
+
+  std::string label = "hcalRawToDigiGPU";
+  confDesc.add(label, desc);
+}
+
+HcalRawToDigiGPU::HcalRawToDigiGPU(const edm::ParameterSet& ps)
+    : rawDataToken_{consumes<FEDRawDataCollection>(ps.getParameter<edm::InputTag>("InputLabel"))},
+      digisF01HEToken_{produces<ProductTypef01>(ps.getParameter<std::string>("digisLabelF01HE"))},
+      digisF5HBToken_{produces<ProductTypef5>(ps.getParameter<std::string>("digisLabelF5HB"))},
+      digisF3HBToken_{produces<ProductTypef3>(ps.getParameter<std::string>("digisLabelF3HB"))},
+      fedsToUnpack_{ps.getParameter<std::vector<int>>("FEDs")} {
+  config_.maxChannelsF01HE = ps.getParameter<uint32_t>("maxChannelsF01HE");
+  config_.maxChannelsF5HB = ps.getParameter<uint32_t>("maxChannelsF5HB");
+  config_.maxChannelsF3HB = ps.getParameter<uint32_t>("maxChannelsF3HB");
+  config_.nsamplesF01HE = ps.getParameter<uint32_t>("nsamplesF01HE");
+  config_.nsamplesF5HB = ps.getParameter<uint32_t>("nsamplesF5HB");
+  config_.nsamplesF3HB = ps.getParameter<uint32_t>("nsamplesF3HB");
+
+  inputCPU_.allocate();
+  inputGPU_.allocate();
+  outputGPU_.allocate(config_);
+  scratchGPU_.allocate(config_);
+  outputCPU_.allocate();
+}
+
+HcalRawToDigiGPU::~HcalRawToDigiGPU() {
+  inputGPU_.deallocate();
```

Comment:
we could simplify the protection in the destructor if we initialise the pointers to `nullptr` when the `CUDAService` is not available, and check for that instead ?

Reply:
> we could simplify the protection in the destructor if we initialise the pointers to `nullptr` when the `CUDAService` is not available, and check for that instead ?

I'd prefer that solution as well.

Reply:
Unfortunately, after looking at `RecoLocalCalo/HcalRecAlgos/interface/DeclsForKernels.h`, `CUDADataFormats/HcalRecHitSoA/interface/RecHitCollection.h` *etc.* it looks like this would require changes deeper in the chain of data types :-(

I'll stick to `if (cs and cs->enabled()) ...` also for the destructors.